### ISSUE TITLE
Unable to Reproduce Megadepth-1500 Evaluation Results

### ISSUE THREAD

**Issue by AbyssGaze** [Parskatt/DeDoDe#12]
I have run the evaluation in Megadepth-1500, but I can't reproduce the results. Some experiments are listed in the table below:
![企业微信截图_1692795533225](https://github.com/Parskatt/DeDoDe/assets/14217667/98b2f175-5308-479e-83cf-80818fd42be0)

 If you could provide some details about the experimental parameters, that would be great. Thank you very much!

**Comment by Parskatt** [comment]
Hi! Yes this does look significantly worse. We ran the sota experiments with 30k keypoints in resized resolution 784, 784 for both dedode-B and -G.

**Comment by Parskatt** [comment]
The time in MS looks very low if it includes the ransac. What softmax temperature did you use?

**Comment by Parskatt** [comment]
I'll have a look tonight and compare with our internal code to see where this issue is coming from.

**Comment by AbyssGaze** [comment]
Thanks for your reimplementation. 
I used the dual_softmax_matcher as in your demo script, and the metric is RANSAC time for pose estimation cost. The time only accounts for the cost in your pipeline, such as detect, describe, and DualSoftMaxMatcher. I only used 10k keypoints for matching. I will test the performance using 30k keypoints later.

**Comment by AbyssGaze** [comment]
def test_step(self, batch, batch_idx):
    with self.profiler.profile("dedode detect"):
        detections_A = self.detector.detect(
            {"image": batch["image0"]}, num_keypoints=30_000
        )
        keypoints_A, P_A = detections_A["keypoints"], detections_A["confidence"]
        detections_B = self.detector.detect(
            {"image": batch["image1"]}, num_keypoints=30_000
        )
        keypoints_B, P_B = detections_B["keypoints"], detections_B["confidence"]
    with self.profiler.profile("dedode describe"):
        description_A = self.descriptor.describe_keypoints(
            {"image": batch["image0"]}, keypoints_A
        )["descriptions"]
        description_B = self.descriptor.describe_keypoints(
            {"image": batch["image1"]}, keypoints_B
        )["descriptions"]
    with self.profiler.profile("dedode match"):
        matches_A, matches_B, batch_ids = self.matcher.match(
            keypoints_A,
            description_A,
            keypoints_B,
            description_B,
            P_A=P_A,
            P_B=P_B,
            normalize=True,
            inv_temp=20,
            threshold=0.01,
        )  # Increasing threshold -> fewer matches, fewer outliers
        W_A, H_A = batch["image0"].shape[3], batch["image0"].shape[2]
        W_B, H_B = batch["image1"].shape[3], batch["image1"].shape[2]
        matches_A, matches_B = self.matcher.to_pixel_coords(
            matches_A, matches_B, H_A, W_A, H_B, W_B
        )
        batch["mkpts0_f"] = matches_A * batch["scale0"]
        batch["mkpts1_f"] = matches_B * batch["scale1"]
        batch["m_bids"] = torch.zeros(matches_A.shape[0]).to(matches_A.device)

**Comment by Parskatt** [comment]
We use a fixed intrinsic of longer length 1200 during eval, with a pixel threshold of 0.5. It might be the case that the version you are using does something else?

**Comment by Parskatt** [comment]
I'm saying this because it seems that the main difference between our results and the ones you report are in AUC@5, and AUC@20 is more similar. This indicates to me that there is a difference in the estimator used.

**Comment by AbyssGaze** [comment]
Ok, thanks for your suggestion.

**Comment by Parskatt** [comment]
Also how did you load the images, are they grayscale in loftr, and are they even normalized? From what I remember the loftr image loading is quite weird, might mess with dedode

**Comment by AbyssGaze** [comment]
have converted grayscale images to color with three channels and used the same augmentation with parameters as: mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]. I have reviewed your evaluation code, and it appears that you scaled the intrinsic parameters (essentially scaling the images) before evaluating. In addition, the RANSAC threshold scale also changes, as shown in the code below:
```
# LoFTR
ransac_thr = thresh / np.mean([K0[0, 0], K1[1, 1], K0[0, 0], K1[1, 1]])
# Dedode
norm_threshold = threshold / (np.mean(np.abs(K1[:2, :2])) + np.mean(np.abs(K2[:2, :2])))
```
I think this may not be entirely fair when comparing with other algorithms. As I used 30k keypoints for matching, the metric increased:
![企业微信截图_1692795533225](https://github.com/Parskatt/DeDoDe/assets/14217667/5a2c5561-9406-4fe5-ac52-c9675ac26b5a)
I will check again to ensure that correct.

**Comment by Parskatt** [comment]
It depends what you mean with fair :D other works are resizing the images and the intrinsics correspondingly, this is quite similar to what we do. We used the same eval for all the other detector descriptor methods, so I think it's fair.

**Comment by Parskatt** [comment]
Yeah and the loftr normalization is bugged, but it doesnt cause major changes.

**Comment by AbyssGaze** [comment]
You may be right, so I am currently testing to ensure that these methods are evaluated under the same conditions. Some experiments are as follows:
![企业微信截图_1692796044158](https://github.com/Parskatt/DeDoDe/assets/14217667/f9bf509a-07fe-4da7-ae9d-4265edebaa6a)
Your paper offers insightful thoughts, and I greatly appreciate your work.

**Comment by Parskatt** [comment]
I'm guessing the ransac threshold you use is for 784 instead of 1200 longer side? This might matter for sharp thresholds. I agree that it might be somewhat unfair to loftr to use 1200, on the other hand aspanformer used 1152, and lightglue use the best performing threshold. Basically its not easy to make it fair for everyone.

**Comment by Parskatt** [comment]
I think the fairest approach is that of lightglue, use the optimal threshold for each matcher. Then probably loftr would have a higher score (pretty sure they should beat us).

**Comment by AbyssGaze** [comment]
Yes, compared to detection and segmentation tasks, our field is more significantly influenced by pre- and post-processing. It seems essential to build a unified codebase to ensure fairness in evaluating various algorithms. In summary, I greatly appreciate your response and hope that we can all contribute to this field together.

**Comment by Parskatt** [comment]
> I have converted grayscale images to color with three channels and used the same augmentation with parameters as: mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]. I have reviewed your evaluation code, and it appears that you scaled the intrinsic parameters (essentially scaling the images) before evaluating. In addition, the RANSAC threshold scale also changes, as shown in the code below:
> 
> ```
> # LoFTR
> ransac_thr = thresh / np.mean([K0[0, 0], K1[1, 1], K0[0, 0], K1[1, 1]])
> # Dedode
> norm_threshold = threshold / (np.mean(np.abs(K1[:2, :2])) + np.mean(np.abs(K2[:2, :2])))
> ```
> 
> I think this may not be entirely fair when comparing with other algorithms. As I used 30k keypoints for matching, the metric increased: ![企业微信截图_1692795533225](https://user-images.githubusercontent.com/14217667/262675804-5a2c5561-9406-4fe5-ac52-c9675ac26b5a.png) I will check again to ensure that correct.

You cannot convert grayscale tocrgb, information will be lost. Although I guess the impact of this is minor.

**Comment by AbyssGaze** [comment]
I read the images as RGB directly, not by converting from grayscale.

**Comment by Parskatt** [comment]
I will try running our own benchmark and compare it with 784,784 intrinsics later tonight.

**Comment by Parskatt** [comment]
Although probably better to compare to loftr with optimal threshold

**Comment by Parskatt** [comment]
I'm not able to reproduce the poor results. I set the intrinsics to be equivalent of longer side 840, and got the following results:

> AUC@5: 58.6, AUC@10: 71.6, AUC@20: 81.0

I'll upload the eval script I used

**Comment by Parskatt** [comment]
See here:

https://github.com/Parskatt/DeDoDe/blob/main/experiments/eval/eval_dedode_descriptor-B.py

Not exactly sure why it works better than previously. @AbyssGaze please let me know if you get similar results (+-1) running this script.

**Comment by AbyssGaze** [comment]
I noticed a significant difference in the results when comparing the ransac method used in your script, which is cv2.USAC_ACCURATE, to the cv2.RANSAC method. When I used cv2.RANSAC without making any other changes, the metric decreased to:
`None auc: [0.4340217424129329, 0.6105990766754759, 0.7489782173192135]`

**Comment by Parskatt** [comment]
Oh wow, there's a double definition of estimate_pose in utils.py... That's very unfortunate. Let me rerun the experiments too.

**Comment by Parskatt** [comment]
@AbyssGaze Yes, unfortunately, I'm able to reproduce your results. I'll rerun the eval on mega-1500 for all keypoint methods (we used the same incorrect estimator for all), and update the paper as soon as possible. Thanks a lot for finding this out, it would have been horrible to find out later!

**Comment by Parskatt** [comment]
Without adjusting the intrinsics (I guess this corresponds to the setting in LoFTR), we get: 

[49.4, 65.5, 77.7]

**Comment by AbyssGaze** [comment]
Thank you for your reply. It is necessary to maintain the same ransac method throughout the entire evaluation. There is also a lot of work being done on subsequent outlier filtering, which plays a crucial role in improving overall performance.

**Comment by Parskatt** [comment]
> It is necessary to maintain the same ransac method throughout the entire evaluation

Yes, definitely!  

> There is also a lot of work being done on subsequent outlier filtering, which plays a crucial role in improving overall performance.

Yes, in this work we wanted to see how good performance we get without any such filtering. Of course we expect the performance of DeDoDe to increase further with filtering.

**Comment by Parskatt** [comment]
I'm closing this and opening a new issue #13 to track the updated results.
